src/content/blog/2023-09-20-tusd-200.md

```diff
@@ -0,0 +1,65 @@
+---
+title: 'tusd v2: More interoperable and resilient'
+author: acconut
+redirect_from: /blog/2023/09/20/tusd-200/
+date: 2023-09-20
+---
+Exactly fours years after the last major release of tusd, [version 1.0.0](https://tus.io/blog/2019/09/20/tusd-100), it is time for another big announcement: We are happy to say that tusd v2 is finished and ready for production!
+
+This major release focuses on two main aspects: **interoperability and resilience**. The reworked hook system makes it easier to integrate tusd into your application, while also extending its functionality. Furthermore, we focused on how tusd handles degraded network connectivity and made numerous changes to improve its resilience in such circumstances. Besides these main aspects, there are additional improvements, which are all described in the following sections.
+
+Before we talk about them in more details, I want to emphasize that there are **no breaking changes in the HTTP interface for tus uploads**. Existing tus clients will continue to be able to communicate with tusd without any changes being necessary. You can replace your tusd v1 deployments with v2 and all clients will continue to function.
+
+## Hook system
+
+This major release includes an entirely reworked hook system. Hooks enable bi-directional communication between the tus upload server and your main application. During the lifecycle of an upload, events are emitted that can notify your application whenever an upload is started, finished, or while data is being transmitted. This allows for a tight integration of resumable uploads into your system.
+
+In the previous versions, this hook interface was mainly design for one-directional communication that flows from tusd to your main application. Over the time, we realized that this approach is too limited and developers want to also send feedback back to tusd in order to reject upload before they are created, send fully-customized error messages to the clients and include additional information in the responses once an upload is complete. tusd v1 did not have proper support for these tasks.
+
+With tusd v2, we have completely reworked the hook system making it more powerful than before:
+
+- Hooks can be sent to your application via HTTP, gRPC, by invoking scripts, or by loading plugins into tusd. Plugins are a new and efficient hook provider, where you can write callbacks using Go and load them into tusd.
+- Hooks can fully customize every HTTP response that is sent from tusd to the tus clients. This includes the status code, header fields, and the response body for success and error cases.
```

Comment:
It should work now with the pre-finish hook as well (next to pre-create and post-receive). Did you test it out and tusd did not behave that way?

Reply:
I thought this was still an issue, but maybe it was fixed in the meantime: https://github.com/tus/tus-js-client/issues/557

Reply:
Ah, ok, I see what you mean. The referenced issue is this unresolved because if pre-finish fails, tusd will still consider the upload done. However, tusd v2 allows you to customize the response from pre-finish (even though the hook will never be retried and the data is lost if the client fails to receive the response). So it is not perfect, but at least a step in the right direction. Does that make sense?